src/lib/cmdgen.c

```diff
@@ -831,66 +820,85 @@ struct cmd_info **lyd2cmds(const struct lyd_node *change_node)
+                    "make sure root container has ipr2cgen:cmd-update\n",
+                    __func__);
+            return NULL;
+        }
 
-    struct lyd_node *next, *child_node, *rollback_dnode;
-    child_node = lyd_child(ordered_change_node);
-    // loop through children of the root node.
-    LY_LIST_FOR(child_node, next)
-    {
-        if (next->schema->nodetype == LYS_LIST) {
-            if (is_startcmd_node(next)) {
-                cmd_line = lyd2cmd_line(next, oper2cmd_prefix);
-                if (cmd_line == NULL) {
-                    fprintf(stderr, "%s: failed to generate ipr2 cmd for node \"%s\" \n", __func__,
-                            next->schema->name);
-                    free_cmds_info(cmds);
-                    return NULL;
-                }
-                // get the rollback node.
-                lyd_diff_reverse_all(next, &rollback_dnode);
-                // the reversed node come with no parent, so we need to set the parent
-                // for the rollback so we can fetch data from sr, otherwise the fetch will fail.
-                rollback_dnode->parent = next->parent;
-                rollback_cmd_line = lyd2cmd_line(rollback_dnode, oper2cmd_prefix);
-                if (rollback_cmd_line == NULL) {
-                    fprintf(stderr, "%s: failed to generate ipr2 rollback cmd for node \"%s\" \n",
-                            __func__, rollback_dnode->schema->name);
-                    free_cmds_info(cmds);
-                    return NULL;
+        struct lyd_node *next, *child_node, *rollback_dnode;
+        child_node = lyd_child(ordered_change_node);
+        // loop through children of the root node.
+        LY_LIST_FOR(child_node, next)
+        {
+            if (next->schema->nodetype == LYS_LIST) {
+                if (is_startcmd_node(next)) {
+                    cmd_line = lyd2cmd_line(next, oper2cmd_prefix);
+                    if (cmd_line == NULL) {
+                        fprintf(stderr, "%s: failed to generate ipr2 cmd for node \"%s\" \n",
+                                __func__, next->schema->name);
+                        free_cmds_info(cmds);
+                        return NULL;
+                    }
+                    // get the rollback node.
+                    lyd_diff_reverse_all(next, &rollback_dnode);
+                    // the reversed node come with no parent, so we need to set the parent
+                    // for the rollback so we can fetch data from sr, otherwise the fetch will fail.
+                    rollback_dnode->parent = next->parent;
+                    rollback_cmd_line = lyd2cmd_line(rollback_dnode, oper2cmd_prefix);
+                    if (rollback_cmd_line == NULL) {
+                        fprintf(stderr,
+                                "%s: failed to generate ipr2 rollback cmd for node \"%s\" \n",
+                                __func__, rollback_dnode->schema->name);
+                        free_cmds_info(cmds);
```

Comment:
need to free lyd_nodes before return:
lyd_free_tree(ordered_change_node);
lyd_free_tree(rollback_dnode);

Reply:
or maybe more elegantly create a `cleanup `code target and `goto cleanup `